rclcpp/include/rclcpp/node_interfaces/node_base.hpp

```diff
@@ -15,27 +15,42 @@
 #ifndef RCLCPP__NODE_INTERFACES__NODE_BASE_HPP_
 #define RCLCPP__NODE_INTERFACES__NODE_BASE_HPP_
 
+#include <atomic>
+#include <functional>
 #include <memory>
+#include <mutex>
 #include <string>
+#include <unordered_map>
 #include <vector>
 
 #include "rcl/node.h"
+#include "rclcpp/callback_group.hpp"
 #include "rclcpp/context.hpp"
 #include "rclcpp/macros.hpp"
 #include "rclcpp/node_interfaces/node_base_interface.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+
 namespace rclcpp
 {
 namespace node_interfaces
 {
 
+class map_of_mutexes;
+
 /// Implementation of the NodeBase part of the Node API.
 class NodeBase : public NodeBaseInterface
 {
 public:
   RCLCPP_SMART_PTR_ALIASES_ONLY(NodeBase)
 
+  static std::unique_ptr<map_of_mutexes> map_object_ptr;
+  static bool map_init_flag;
+  static std::mutex map_init_flag_mutex;
```

Comment:
do we need to lazely initialize the map?
If yes, you can use std::call_once() and std::once_flag instead of doing it manually.

Reply:
I don't think we need to lazely initialize the map, it seems to me that it can be initialized before main as all other globals.

Reply:
Yeah, I agree.  I think that will be better, and will also get rid of the need for the extra mutex while lazily initializing.